avatar/metrics/trace.py

```diff
@@ -0,0 +1,241 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Avatar metrics trace."""
+
+import re
+import time
+import types
+
+from avatar.metrics.trace_pb2 import (
+    DebugAnnotation,
+    ProcessDescriptor,
+    ThreadDescriptor,
+    Trace,
+    TracePacket,
+    TrackDescriptor,
+    TrackEvent,
+)
+from google.protobuf.text_encoding import CUnescape
+from mobly.base_test import BaseTestClass
+from typing import TYPE_CHECKING, Any, List, Protocol, Union
+
+if TYPE_CHECKING:
+    from avatar import PandoraDevices
+    from avatar.pandora_client import PandoraClient
+else:
+    PandoraClient = object
+    PandoraDevices = object
+
+
+packets: List[TracePacket] = []
+genesis: int = time.monotonic_ns()
+uuid: int = 0
+
+
+def process_name(test: BaseTestClass) -> str:
+    return f"{test.__class__.__name__}.{test.current_test_info.name}"
+
+
+def next_uuid() -> int:
+    global uuid
+    uuid += 1
+    return uuid
+
+
+def hook_test(test: BaseTestClass) -> None:
+    global packets
+    original_teardown_class = test.__class__.teardown_class
+    original_setup_test = test.__class__.setup_test
+
+    def teardown_class(self: BaseTestClass) -> None:
+        output_path: str = test.current_test_info.output_path  # type: ignore
+        trace = Trace(packet=packets)
+        with open(f"{output_path}/avatar.trace", "wb") as f:
+            f.write(trace.SerializeToString())
+        with open(f"{output_path}/packets.log", "a") as f:
+            for packet in packets:
+                f.write(f"{packet}")
+                f.write("----------\n")
+
+        original_teardown_class(self)
+
+    def setup_test(self: BaseTestClass) -> None:
+        global genesis
+        genesis = time.monotonic_ns()
+        assert hasattr(self, "devices")
+        process_id = next_uuid()
+        devices: PandoraDevices = getattr(self, "devices", [])  # type: ignore
+        packets.append(
+            TracePacket(
+                track_descriptor=TrackDescriptor(
+                    uuid=process_id,
+                    process=ProcessDescriptor(pid=process_id, process_name=process_name(self)),
```

Comment:
`process_name` only seems to be used here an is a one line function can we inline it ?